model/jpa/src/main/java/org/keycloak/storage/certificate/jpa/JpaCertificateStorageProvider.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2025 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.storage.certificate.jpa;
+
+import java.security.KeyPair;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.LockModeType;
+import org.keycloak.common.util.PemUtils;
+import org.keycloak.storage.certificate.CertificateDetail;
+import org.keycloak.storage.certificate.CertificateStorageProvider;
+import org.keycloak.storage.certificate.jpa.entity.CertificateDetailEntity;
+
+public class JpaCertificateStorageProvider implements CertificateStorageProvider {
+
+    private final EntityManager entityManager;
+
+    public JpaCertificateStorageProvider(EntityManager entityManager) {
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public CertificateDetail load(String id) {
+        var entity = entityManager.find(CertificateDetailEntity.class, Objects.requireNonNull(id));
+        if (entity == null) {
+            return null;
+        }
+        return entityToDetails(entity);
+    }
+
+    @Override
+    public void store(String id, CertificateDetail certificateDetail) {
+        var entity = detailsToEntity(id, Objects.requireNonNull(certificateDetail));
+        entityManager.persist(entity);
+    }
+
+    @Override
+    public void remove(String id) {
+        var entity = entityManager.find(CertificateDetailEntity.class, id, LockModeType.PESSIMISTIC_WRITE);
+        if (entity != null) {
+            entityManager.remove(entity);
+        }
+    }
+
+    @Override
+    public CertificateDetail loadOrCreate(String id, Supplier<CertificateDetail> generator) {
+        var entity = entityManager.find(CertificateDetailEntity.class, id, LockModeType.PESSIMISTIC_WRITE);
+        if (entity != null) {
```

Comment:
I know we don't to rotations yet, still could we at least create a new certificate one if we see that the existing one has expired? Either on the same keys, or on new keys, whatever is simpler for us?

Reply:
I would leave this for the follow-up PR. It has to be done outside the storage otherwise, it can keep the Keycloak server outside of the cluster. Any change needs to have an event for existing members to rotate the certificate. 

Also, the certificate returned by `CertificateUtilsProvider#generateV1SelfSignedCertificate(java.security.KeyPair, java.lang.String)` has a 10-year expiration. I'm planning to finish the rotation in 26.2, and not take 10 years :smile:

Reply:
OK, let's proceed without for now. For every release, there are people running it for years unchanged, so please have the rotation in place for 26.2. Sad, not funny.